File: BinarySearch/C++/1011-Capacity-To-Ship-Packages-Within-D-Days.cpp
class Solution {
public:
    int calculateSum(vector<int>& weights, int D, int cap){
        int i = 0;
        for(int day = 0; day < D; day++){
            int maxCap = cap;
            while((maxCap -= weights[i]) >= 0){
                i++;
                if(i == weights.size()){
                    return true;
                }
            }
        }
        return false;
    }
    int shipWithinDays(vector<int>& weights, int D) {
        int right = 0;//右边界
        int left = 1;
        int sum = 0;
        for(auto weight : weights){
            sum += weight;
        }
        right += sum;
        while(left < right){
            int mid = left + (right - left) / 2;
            if(calculateSum(weights, D, mid)){
                right = mid;   
            }
            else{
                left = mid + 1;
            }
        }
        return left;
    }
};

File: BinarySearch/Go/1011. Capacity To Ship Packages Within D Days.go
func shipWithinDays(weights []int, D int) int {
	left, right := 1, 0
	for i := 0; i < len(weights); i++ {
		right += weights[i]
	}

	for left < right {
		mid := (left + right) >> 1
		if helper(weights, D, mid) {
			right = mid
		} else {
			left = mid + 1
		}
	}
	return left
}

func helper(weights []int, D int, cap int) bool {
	i := 0
	cur := 0
	days := 0

	for i < len(weights) {
		for i < len(weights) && cur+weights[i] <= cap {
			cur += weights[i]
			i++
		}

		if cur == 0 {
			return false
		} else {
			days++
			cur = 0
		}
	}

	if days <= D {
		return true
	} else {
		return false
	}
}